crates/iceberg/src/expr/visitors/bound_predicate_visitor.rs

```diff
@@ -0,0 +1,366 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::expr::{BoundPredicate, BoundReference, PredicateOperator};
+use crate::spec::Datum;
+use crate::Result;
+use fnv::FnvHashSet;
+
+pub trait BoundPredicateVisitor {
+    type T;
+
+    fn always_true(&mut self) -> Result<Self::T>;
+    fn always_false(&mut self) -> Result<Self::T>;
+
+    fn and(&mut self, lhs: Self::T, rhs: Self::T) -> Result<Self::T>;
+    fn or(&mut self, lhs: Self::T, rhs: Self::T) -> Result<Self::T>;
+    fn not(&mut self, inner: Self::T) -> Result<Self::T>;
+
+    fn is_null(&mut self, reference: &BoundReference) -> Result<Self::T>;
```

Comment:
The problem with this approach is that when we add an operator, we need to modify the trait. We can still keep it compatible with default function, but the actual behavior may break when user upgrades library. cc @viirya @Xuanwo @marvinlanhenke What do you think?

Reply:
If we add default implementations that return something like `Err(OperatorNotImplemented)`, then if we add an operator, wouldn't existing code work for existing operators and only break if users then try to use the new operator without having updated any visitors they have to use the new operator?

Reply:
> If we add default implementations that return something like Err(OperatorNotImplemented), then if we add an operator, wouldn't existing code work for existing operators and only break if users then try to use the new operator without having updated any visitors they have to use the new operator?

Yes, I agree with that.

Another option is to use following method:
```rust
trait BoundPredicateVisitor {
  fn visitor_op(&mut self, op: &PredicateOperator, reference: &BoundReference, results: Vec<Self::T>) -> Result<Self::T> {}
}
```

This way we can force user to think about added new operator, otherwise it will throw compiler error, but this may break things when we add new operator.

I think the two approaches have different pros and cons, let's wait to see others's comments.

Reply:
> Another option is to use following method:
> 
> ```rust
> trait BoundPredicateVisitor {
>   fn visitor_op(&mut self, op: &PredicateOperator, reference: &BoundReference, results: Vec<Self::T>) -> Result<Self::T> {}
> }
> ```

I think I'm in favor of keeping the trait as generic as possible - so we don't have to modify it, if we have to support a new operator. 

When implementing this trait (no matter which variant) I'd most likely have to overwrite the default behaviour anyway? So I wouldn't bother providing any in the first place and keep the trait as generic and future proof as possible? If that makes any sense to you?

Reply:
> ...wouldn't I have match op anyway, so having a default match arm in the implementation would solve the issue for the downstream user?

I think this is the advantage of the op approach, it's left to user to decide what to do. If they want new op to be ignored, they could use a default match arm, otherwise they could choose to do pattern match against each operator.

Reply:
> I think this is the advantage of the op approach, it's left to user to decide what to do. If they want new op to be ignored, they could use a default match arm, otherwise they could choose to do pattern match against each operator.

Yes, exactly. So I'm +1 for the op approach.

Reply:
I don't see how this will work with just `visitor_op`.

The proposed function signature `fn visitor_op(&mut self, op: &PredicateOperator, reference: &BoundReference)` would work for a unary operator.

A binary operator would need `fn visitor_op(&mut self, op: &PredicateOperator, reference: &BoundReference, literal: &Datum)`, and a set operator would need `fn visitor_op(&mut self, op: &PredicateOperator, reference: &BoundReference, literals: &FnvHashSet<Datum>)`.

To resolve this, you'd need the visitor to have `fn unary_op(...)`, `fn binary_op(...)`, and `fn set_op(...)`.

Is that what we want? 

Personally I think this is another example of where the current type hierarchy feels awkward - there are other places where we bump into these type hierarchy mismatches, such as needing a default leg in a match on operator to handle operators that will never show up, such as binary ops in a unary expression.

Something more like this feels better to me:

```rust
pub enum Predicate {
    // By not having Unary / Binary / Set here, and removing
    // that layer in favour of having the ops themselves
    // as part of this enum, there is no need for code that
    // raises errors when the wrong type of `Op` is used, since those
    // issues are prevented by the type system at compile time

    // Unary ops now only contain one param, the term, and so can be
    // implemented more simply
    IsNull(Reference),
    // ... other Unary ops here,

    // Binary / Set ops ditch the `BinaryExpression`, with the literal
    // and reference being available directly
    GreaterThan {
        literal: Datum,
        reference: Reference,
    },
    // ... other Binary ops here,
    
    In {
        literals: HashSet<Datum>,
        reference: Reference,
    },
    // ... other Set ops here
}
```

Reply:
I also like the idea of going further and having an `Expression` enum that contains the logical operators plus `Predicate` above, with `Reference` from the above being an enum over bound or unbound. This allows visitors to potentially work over both bound and unbound expressions.

```rust
pub enum Expression {
    // The top level would be Expression rather than Predicate.
    // The logical operators would live at this level.

    // Since logical operators such as And / Or / Not
    // are common between bound and unbound, some visitors
    // such as rewrite_not can be applied to both bound and
    // unbound expressions, preventing duplication and increasing
    // flexibility

    /// An expression always evaluates to true.
    AlwaysTrue,

    /// An expression always evaluates to false.
    AlwaysFalse,

    // Here we use a struct variant to eliminate the need
    // for a generic `LogicalExpression`, which did not
    // add much, also allowing us to refer to lhs and rhs
    // by name

    /// An expression combined by `AND`, for example, `a > 10 AND b < 20`.
    And {
        lhs: Box<Expression>,
        rhs: Box<Expression>
    },

    /// An expression combined by `OR`, for example, `a > 10 OR b < 20`.
    Or {
        lhs: Box<Expression>,
        rhs: Box<Expression>
    },

    /// An expression combined by `NOT`, for example, `NOT (a > 10)`.
    Not(Box<Expression>),

    /// A Predicate
    Predicate(Predicate),
}

pub enum Reference {
    // Only at the leaf level do we distinguish between
    // bound and unbound. This allows things like bind
    // and unbind to be implemented as a transforming visitor,
    // and avoids duplication of logic compared to having
    // Bound / Unbound being distinguished at a higher level
    // in the type hierarchy

    Bound(BoundReference),
    Unbound(UnboundReference),
}

pub struct UnboundReference {
    name: String,
}

pub struct BoundReference {
    column_name: String,
    field: NestedFieldRef,
    accessor: StructAccessor,
}
``` 

We could use type alises such as `type BoundExpression = Expression` and have `bind` return a `BoundExpression` to prevent bound expressions being used where unbound ones are required and vice versa. But I'm digressing very far from the original comment now, sorry! 😆

Reply:
> The proposed function signature fn visitor_op(&mut self, op: &PredicateOperator, reference: &BoundReference) would work for a unary operator.

I think following signature would be applied to all operators:
```rust
 fn visitor_op(&mut self, op: &PredicateOperator, reference: &BoundReference, literals: &[Datum])
```

Reply:
> I also like the idea of going further and having an Expression enum that contains the logical operators plus Predicate above, with Reference from the above being an enum over bound or unbound. This allows visitors to potentially work over both bound and unbound expressions.

Hi, @sdd Would you mind to open an issue to raise a discussion about this idea? I think it looks promsing to me.

Reply:
That still wouldn't work without collecting the set into a vec first for set ops. I've updated the PR with an alternative based on an enum, see what you think.